Author: Ken Zhu <quic_guigenz@quicinc.com>
Date:   Fri Jan 3 16:57:07 2025 -0800

    firewall4: add connlimit_ports support

    interpret the connlimit_ports to nft commands.

    Change-Id: I4d185ea4ff945f1f03283861fc0a9a48f355092b
    Signed-off-by: Ken Zhu <quic_guigenz@quicinc.com>

diff -Nur a/root/usr/share/firewall4/templates/redirect.uc b/root/usr/share/firewall4/templates/redirect.uc
--- a/root/usr/share/firewall4/templates/redirect.uc	2024-05-20 23:54:02.000000000 -0700
+++ b/root/usr/share/firewall4/templates/redirect.uc	2025-01-03 16:45:34.249083274 -0800
@@ -72,6 +72,10 @@
 	counter {%+ endif -%}
 {%+  if (redirect.log): -%}
 	log prefix {{ fw4.quote(redirect.log, true) }} {%+ endif -%}
+
+{%+  if (redirect.count): -%}
+	ct count {{ redirect.count }} {%+ endif -%}
+
 {%   if (redirect.target == "redirect"): -%}
 	redirect{% if (redirect.rport): %} to {{ fw4.port(redirect.rport) }}{% endif %}
 {%-  elif (redirect.target == "accept" || redirect.target == "masquerade"): -%}
diff -Nur a/root/usr/share/ucode/fw4.uc b/root/usr/share/ucode/fw4.uc
--- a/root/usr/share/ucode/fw4.uc	2025-01-03 16:41:00.995188746 -0800
+++ b/root/usr/share/ucode/fw4.uc	2025-01-03 16:52:39.543985838 -0800
@@ -2995,6 +2995,7 @@
 	},
 
 	parse_nat: function(data) {
+		let count = 0;
 		let snat = this.parse_options(data, {
 			enabled: [ "bool", "1" ],
 
@@ -3072,6 +3073,10 @@
 			return;
 		}
 
+		if (snat.connlimit_ports) {
+			this.count = snat.snat_port.max - snat.snat_port.min +1;
+		}
+
 		switch (this.parse_bool(snat.log)) {
 		case true:
 			snat.log = `${snat.name}: `;
@@ -3102,6 +3107,7 @@
 
 				raddr: raddrs ? raddrs[0] : null,
 				rport: rport,
+				count: this.count,
 
 				chain: snat.src?.zone ? `srcnat_${snat.src.zone.name}` : "srcnat"
 			};
